app/Http/Controllers/RecipeController.php

```diff
@@ -12,54 +14,211 @@ class RecipeController extends Controller
+                'description' => $request->description,
+                'instructions' => $request->instructions,
+                'prep_time' => $request->prep_time,
+                'cook_time' => $request->cook_time,
+                'total_time' => $request->total_time,
+                'servings' => $request->servings,
+                'image_url' => $request->image_url,
+                'ingredients' => collect(explode('\\n', $request->ingredients))->toJson(),
+                'nutritional_info' => collect(explode('\\n', $request->nutritional_info))->toJson(),
+            ]);
+
+            // Assign tags to recipe
+            if ($request->has('tags')) {
+                // Delete previous tags
+                foreach ($recipe->tags as $tag) {
+                    $tag->delete();
+                }
+
+                // Recreate new tags
+                foreach ($request->tags as $tag) {
+                    $recipe->tags()->create(
+                        ['tag' => $tag]
+                    );
+                }
+            }
+
+            return redirect()->route('recipe.show', $recipe)->with('success', 'Recipe updated');
+        } catch (\Throwable $th) {
+            return redirect()->back()->with('error', $th->getMessage());
+        }
     }
 
     /**
      * Remove the specified resource from storage.
      */
     public function destroy(Recipe $recipe)
     {
-        //
+        // Todo: Will use events to delete this
+    }
+
+    /**
+     * Add or remove collaborators from recipes
+     */
+    public function add_collaborators(Request $request)
+    {
+        // Todo: Work on collaborators
+
+        // Validate Recipe fields
+        // $request->validate([
+        //     $request->collaborators => 'string|required'
+        // ]);
+
+        // // Assign Collaborators to Recipe
+
+        // // $validator = Validator::make($request->all(),  [
+
+        // foreach(explode(';', $request->collaborators) as $collaborator){ // Expect a string of semicolon delimited email addresses
+        //     $user = User::where('email', $collaborator)->first();
+
+        //     if(!$user){
+
+        //     }
+        // }
+
+    }
+
+    /**
+     * Add a rating to a recipe
+     */
+    public function rate(Request $request, Recipe $recipe)
+    {
+        try {
+            $request->validate([
+                'rating' => 'required|integer|min:1|max:5',
+            ]);
+
+            // To get auth user
+            $recipe->user_ratings()->attach(User::First(), ['rating' => $request->rating]);
+
+            return redirect()->route('recipe.show', $recipe)->with(['success', 'Recipe rated!']);
+        } catch (\Throwable $th) {
+            return redirect()->back()->with('error', $th->getMessage());
+        }
+    }
+
+    public function like(Request $request, Recipe $recipe)
+    {
+        try {
+            // To get auth user
+            $recipe->users_liked()->attach(User::First());
```

Comment:
This needs to be sync instead of attach.

Reply:
Sync will delete all data in the table. Not just the related data. Updated to

```php
// To get auth user
$recipe->users_liked() // Delete all previous ratings
$recipe->users_liked()->attach(User::First());
```